### PR TITLE
nixos/tayga: init

### PR DESCRIPTION
**Pull request by hax404**
###### Description of changes
Add module and test. Module comes from https://github.com/NixOS/nixpkgs/pull/129390

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/networking/tayga.nix [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    dynamic-pool ${strAddr cfg.ipv4.pool}
+    data-dir ${cfg.dataDir}
+  '';
+
+  makeRouteCommands = oper: ''
+    ip addr ${oper} ${cfg.ipv4.router.address} dev ${cfg.tunDevice}
+    ip route ${oper} ${strAddr cfg.ipv4.pool} dev ${cfg.tunDevice}
+    ip addr ${oper} ${cfg.ipv6.router.address} dev ${cfg.tunDevice}
+    ip route ${oper} ${strAddr cfg.ipv6.pool} dev ${cfg.tunDevice}
+  '';
```
Comment (NickCao) [inline nixos/modules/services/networking/tayga.nix:26]: Could we use `networking` options to implement this instead of ad-hoc iproute2 command?

Comment (mweinelt) [reply nixos/modules/services/networking/tayga.nix:26]: I already talked to hax404 about this and proposed to use networkd units instead. I'm opposed to relying on the highly undermaintained scripted networking parts of nixos.

File: nixos/modules/services/networking/tayga.nix [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+      tunDevice = mkOption {
+        type = types.str;
+        default = "nat64";
+        description = lib.mdDoc "Name of the nat64 tun device";
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd.network = {
```
Comment (NickCao) [inline nixos/modules/services/networking/tayga.nix:123]: `networking.interfaces` should be used as it is backend agnostic and more future proof.

File: nixos/modules/services/networking/tayga.nix [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+      useNetworkd = mkOption {
+        type = types.bool;
+        default = false;
+        description = lib.mdDoc "Use systemd-networkd to configure the nat64 tun device";
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    #networking.interfaces."${cfg.tunDevice}" = mkIf (!config.systemd.network.enable) {
```
Comment (hax404) [inline nixos/modules/services/networking/tayga.nix:129]: I would like to check if systemd-networkd is enabled as on other lines. But `mkIf (!config.systemd.network.enable)` runs into infinitie recursion.
```
georg@rick:~/git/nixpkgs/ > nom-build -j0 . -A nixosTests.tayga
/nix/store/23yy920js0pk5924c3899ajf7kxw7i7y-vm-test-run-tayga
Finished at 21:49:58 after 13s
georg@rick:~/git/nixpkgs/ > nom-build -j0 . -A nixosTests.tayga
error: infinite recursion encountered

       at /home/georg/git/nixpkgs/lib/modules.nix:528:12:

          527|         zipAttrsWith (n: concatLists)
          528|           (map (module: let subtree = module.${attr}; in
             |            ^
          529|               if !(builtins.isAttrs subtree) then
(use '--show-trace' to show detailed location information)
Finished at 21:50:06 after 1s
```

**Comment by NickCao** [comment]
I'm still confused by the useNetworkd part, why can't we keep only the generic implementation.

**Comment by hax404** [comment]
Now I use the generic implementation. To get the most of the sandboxing, there is a check if systemd-networkd is used, because `virtualOwner` cannot be empty in the nixos-scripts and then uses `root` by defaut.

File: nixos/modules/services/networking/tayga.nix [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+        ];
+        CapabilityBoundingSet = "";
+        RestrictAddressFamilies = [
+          "AF_INET"
+          "AF_INET6"
+          "AF_NETLINK"
+        ];
+        StateDirectory = "tayga";
+        DynamicUser = mkIf config.networking.useNetworkd true;
+        User = mkIf config.networking.useNetworkd "tayga";
```
Comment (NickCao) [inline nixos/modules/services/networking/tayga.nix:177]: User name should default to unit name, no need to set it.

File: pkgs/tools/networking/tayga/default.nix [hunk 5]
```diff
@@ -9,6 +9,8 @@ stdenv.mkDerivation rec {
     sha256 = "1700y121lhvpna49bjpssb7jq1abj9qw5wxgjn8gzp6jm4kpj7rb";
   };
 
+  passthru.tests.teeworlds = nixosTests.tayga;
```
Comment (NickCao) [inline pkgs/tools/networking/tayga/default.nix:12]: Is `teeworlds` is typo?

File: nixos/tests/all-tests.nix [hunk 6]
```diff
@@ -620,6 +620,7 @@ in {
   systemd-misc = handleTest ./systemd-misc.nix {};
   tandoor-recipes = handleTest ./tandoor-recipes.nix {};
   taskserver = handleTest ./taskserver.nix {};
+  tayga = handleTest ./tayga.nix {};
```
Comment (NickCao) [inline nixos/tests/all-tests.nix:623]: Better to test both the legacy and networkd implementation.

**Review by NickCao — APPROVED** [review]
Apart from the comments, LGTM

File: nixos/modules/services/networking/tayga.nix (hunk 4 above)
Comment (hax404) [reply nixos/modules/services/networking/tayga.nix]: thanks

File: pkgs/tools/networking/tayga/default.nix (hunk 5 above)
Comment (hax404) [reply pkgs/tools/networking/tayga/default.nix]: oh yes, thanks

File: nixos/tests/all-tests.nix (hunk 6 above)
Comment (hax404) [reply nixos/tests/all-tests.nix:623]: Inside the test is done with both the legacy and systemd based network config. That is a small loop in line 251 that tests the routers after each other.

**Comment by RaitoBezarius** [comment]
Could you move the release notes for 23.05? I am interested into getting this merged.

**Comment by mweinelt** [comment]
We're backporting this into 22.11 since it is unlikely to cause regressions.

**Review by RaitoBezarius — APPROVED** [review]
Looks good to me.

**Review by mweinelt — APPROVED** [review]

**Pull request merged**